### ISSUE TITLE
`git bz apply` fails to detect patch format

### ISSUE THREAD

**Issue by jankeromnes** [mozilla/moz-git-tools#52]
It seems to always fail to detect the patch format lately:

    $ git bz apply 1218817
    Debug: searching for Bugzilla cookies in Firefox profile
    
    Bug 1218817 - Extend the debugger protocol to list all service worker registrations.
    
    patch
    Apply? [yn] y
    
    Patch format detection failed.
    Patch left in /tmp/patch-v7nIPq.patch

But then:

    $ git apply /tmp/patch-v7nIPq.patch
    # success

**Comment by jankeromnes** [comment]
I still hit this regularly. Newest occurrence:

    $ git bz apply 1209559
    Your bugzilla username is not specified in your git config.
    Set it by running:
      git config --global bz.username <your username>
    
    No bugzilla API key is specified in your git config. This is the
    preferred means of authentication, and other methods may break at
    any time. An API key can be obtained here:
      https://bugzilla.mozilla.org/userprefs.cgi?tab=apikey
    Once obtained, set the API key by running:
      git config --global bz.apikey <your bugzilla API key>
    
    Debug: searching for Bugzilla cookies in Firefox profile
    
    Bug 1209559 - Figure out why opening a toolbox on certain workers works only once.
    
    When detaching from a WorkerClient, unregister its ThreadClient.
    Apply? [yn] y
    
    Patch format detection failed.
    Patch left in /tmp/When-detaching-from-a-WorkerClient-unregister-its--uMUllz.patch

The patch format is a straight-up patch, so its format should be too hard to detect.

**Comment by amccreight** [comment]
I can reproduce that error. It seems to be an error from an actual git command. I'll try to take a look at how we are actually applying these patches.

**Comment by jankeromnes** [comment]
Thanks @amccreight !

**Comment by amccreight** [comment]
The issue is that `git bz apply` actually uses `git am` to apply the patch, which expects a nicely formatted patch, but the attachment in bug 1209559 just a raw diff. Probably the best solution here is to change `git bz apply` to use `git apply`, and add a new `git bz am` command that uses `git am`. That will make the behavior less surprising, though it may cause confusion for existing users of `git bz apply`.

**Comment by amccreight** [comment]
Oh, and this should actually be filed against git-bz-moz. I'll file a new issue there.

**Comment by jankeromnes** [comment]
Good idea! I've never really used `git am` and don't know what benefits it has over `git apply`, but I've had to do the occasional `git am --abort` when `git bz` failed.

Thanks for forwarding the issue!

**Comment by amccreight** [comment]
git am does things like set the commit message and author based on the patch info. git apply just modifies the files. In your particular case, the issue is that it was a simple diff without any of the extra info, which git am apparently refuses to deal with.
